data/appRateLimiter.go

```diff
@@ -0,0 +1,305 @@
+		if rate > int64(r.serviceRatePerWindow) {
+			return true
+		}
+		entry.cur.Add(1)
+	}
+
+	return false
+}
+
+func (r *appRateLimiter) len() int {
+	var count int
+	for i := 0; i < numBuckets; i++ {
+		r.buckets[i].mu.RLock()
+		count += len(r.buckets[i].entries)
+		r.buckets[i].mu.RUnlock()
+	}
+	return count
+}
+
+var appKeyPool = sync.Pool{
+	New: func() interface{} {
+		return &appKeyBuf{
+			// max config.MaxTxGroupSize apps per txgroup, each app has up to MaxAppTxnForeignApps extra foreign apps
+			// at moment of writing config.MaxTxGroupSize = 16, config.MaxAppTxnForeignApps = 8
+			keys:    make([]keyType, 0, config.MaxTxGroupSize*(1+config.MaxAppTxnForeignApps)),
+			buckets: make([]int, 0, config.MaxTxGroupSize*(1+config.MaxAppTxnForeignApps)),
+		}
+	},
+}
+
+// appKeyBuf is a reusable storage for key and bucket slices
+type appKeyBuf struct {
+	keys    []keyType
+	buckets []int
+}
+
+func getAppKeyBuf() *appKeyBuf {
+	buf := appKeyPool.Get().(*appKeyBuf)
+	buf.buckets = buf.buckets[:0]
+	buf.keys = buf.keys[:0]
+	return buf
+}
+
+func putAppKeyBuf(buf *appKeyBuf) {
+	appKeyPool.Put(buf)
+}
+
+// txgroupToKeys converts txgroup data to keys
+func txgroupToKeys(txgroup []transactions.SignedTxn, origin []byte, seed uint64, salt [16]byte, numBuckets int) *appKeyBuf {
+	keysBuckets := getAppKeyBuf()
+	// since blake2 is a crypto hash function it seems OK to shrink 32 bytes digest down to 8.
+	// Rationale: we expect thousands of apps sent from thousands of peers,
+	// so required millions of unique pairs => 8 bytes should be enough.
+	// The 16 bytes salt makes it harder to find collisions if an adversary attempts to censor
+	// some app by finding a collision with some app and flood a network with such transactions:
+	// h(app + relay_ip) = h(app2 + relay_ip).
+	var buf [8 + 16 + 16]byte // uint64 + 16 bytes of salt + up to 16 bytes of address
+	txnToDigest := func(appIdx basics.AppIndex) keyType {
+		binary.LittleEndian.PutUint64(buf[:8], uint64(appIdx))
+		copy(buf[8:], salt[:])
+		copied := copy(buf[8+16:], origin)
+
+		h := blake2b.Sum256(buf[:8+16+copied])
+		var key keyType
+		copy(key[:], h[:len(key)])
+		return key
+	}
+	txnToBucket := func(appIdx basics.AppIndex) int {
+		return int(memhash64(uint64(appIdx), seed) % uint64(numBuckets))
+	}
+	for i := range txgroup {
+		if txgroup[i].Txn.Type == protocol.ApplicationCallTx {
+			appIdx := txgroup[i].Txn.ApplicationID
+			// hash appIdx into a bucket, do not use modulo without hashing first since it could
+			// assign two vanilla (and presumable, popular) apps to the same bucket.
+			keysBuckets.buckets = append(keysBuckets.buckets, txnToBucket(appIdx))
+			keysBuckets.keys = append(keysBuckets.keys, txnToDigest(appIdx))
+			if len(txgroup[i].Txn.ForeignApps) > 0 {
+				for _, appIdx := range txgroup[i].Txn.ForeignApps {
+					keysBuckets.buckets = append(keysBuckets.buckets, txnToBucket(appIdx))
+					keysBuckets.keys = append(keysBuckets.keys, txnToDigest(appIdx))
```

Comment:
Foreign app is the only way to assume/expect app call, and this might be under estimation to count it once (like a single foreign app can be called multiple times).

Reply:
After more thought, I think it could make more sense to only count unique apps that appear in a transaction group. Here's my reasoning:
* The situation we want to avoid is people flooding the network with many calls to a single app which will race -- only a very small amount of the app calls will be accepted, while the vast majority will either be rejected or invalid for some other reason.
* App calls in the same transaction group, as well as inner transactions, are all atomically linked. In a flooding/race situation, there's no advantage to packing a larger group or issuing more inners, since if any transaction fails, the entire group fails.
* If we want to throttle apps which are experiencing flood/race events it's therefore sufficient to measure only the unique apps in an atomic group -- whether that group calls the same app 1 time or 16, it shouldn't make a difference. By not measure the unique apps, we are giving additional weight to some types of app interactions which don't necessarily deserve to be throttled to a higher extent.

If you agree, great, but if not, I still think this PR is beneficial as-is, so I'm still prepared to approve (just looking through most recent changes now).

Reply:
not counting foreign apps per txn in a group would allow 16x more calls per some app repeatedly mentioned in a group.

Reply:
implemented